### PR TITLE
Implement implicit rejection for RSA PKCS#1 v1.5 de-padding

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

File: usr/lib/ep11_stdll/new_host.c [hunk 1]
```diff
@@ -2513,17 +2515,27 @@ CK_RV SC_Decrypt(STDLL_TokData_t *tokdata, ST_SESSION_HANDLE *sSession,
                                     length_only, sess->decr_ctx.key,
                                     pEncryptedData, ulEncryptedDataLen,
                                     pData, pulDataLen);
-        if (!is_rsa_mechanism(sess->decr_ctx.mech.mechanism) && rc != CKR_OK)
+        mask = ~constant_time_is_zero(
```
Comment (jschmidb) [inline usr/lib/ep11_stdll/new_host.c:2518]: We could leave the original if statement in comments here, as you did it above. It makes reading the code easier.

File: usr/lib/common/mech_openssl.c [hunk 2]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+    }
+
+    if (BN_bn2binpad(d, priv_exp, priv_exp_len) <= 0) {
+        TRACE_ERROR("BN_bn2binpad failed\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+done:
+    BN_CTX_free(bn_ctx);
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5112]: Interesting: looks like there is no BN_CTX_clear_free, similar to BIGNUMs ...

File: usr/lib/common/mech_openssl.c [hunk 3]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+                            priv_exp_attr->ulValueLen - modulus->ulValueLen;
+        }
+    }
+
+    /*
+     * we use hardcoded hash so that migrating between versions that use
+     * different hash doesn't provide a Bleichenbacher oracle:
+     * if the attacker can see that different versions return different
+     * messages for the same ciphertext, they'll know that the message is
+     * syntethically generated, which means that the padding check failed
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5191]: synthetically

File: testcases/crypto/rsa_func.c [hunk 4]
```diff
@@ -1963,6 +1963,164 @@ CK_RV do_VerifyRSA(struct PUBLISHED_TEST_SUITE_INFO * tsuite)
...
+            goto testcase_cleanup;
+        }
+        // begin testcase
+        testcase_begin("%s Implicit Rejection with test vector %u."
+                       "\npubl_exp='%s', modbits=%lu, publ_exp_len=%lu.",
+                       tsuite->name, i, s,
+                       tsuite->tv[i].mod_len * 8,
+                       tsuite->tv[i].pubexp_len);
+
+        rc = CKR_OK;            // set rc
```
Comment (jschmidb) [inline testcases/crypto/rsa_func.c:2020]: so much space here before the comment?

File: usr/lib/common/mech_openssl.c [hunk 5]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, q) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime2\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(q, BN_FLG_CONSTTIME);
+
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL) {
+        TRACE_ERROR("BN_CTX_get failed d\n");
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5085]: There's a 'd' at the end of the trace string

Comment (jschmidb) [reply usr/lib/common/mech_openssl.c:5085]: Ah ... should be like "BN_CTX_get failed to get d"

File: usr/lib/common/mech_openssl.c [hunk 6]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+    md = EVP_sha256();
+    if (md == NULL) {
+        TRACE_ERROR("EVP_sha256 failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_Digest(priv_exp, modulus->ulValueLen, d_hash, NULL,
+                   md, NULL) <= 0) {
+        TRACE_ERROR("EVP_Digest failed\n");
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5202]: Just a general thing: at other places in opencryptoki, we print out openssl errors with something like ERR_print_errors_cb(openssl_err_cb, NULL); Would this help here also?

File: usr/lib/ep11_stdll/new_host.c (hunk 1 above)
Comment (ifranzki) [reply usr/lib/ep11_stdll/new_host.c:2518]: Ah yes. I did so in the other new_host files, but not here. Will fix.

File: usr/lib/common/mech_openssl.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5112]: Thats the beauty of a BN-context. If you free the BN-context, it automatically frees all BIGNUMs that you obtained from it (via BN_CTX_get()). And BN_CTX_free() also does do a BN_CTX_clear_free() internally on all its BIGNUMs. 
BTW: This here is a secure BN-context (BN_CTX_secure_new()), so all BIGNUMs a resecure BIGNUMs.

File: testcases/crypto/rsa_func.c (hunk 4 above)
Comment (ifranzki) [reply testcases/crypto/rsa_func.c:2020]: We have that same code at other places in this file as well. I just copied it as is.
Will remove the comment here anyway, its just telling the obvious anyway.

File: usr/lib/common/mech_openssl.c (hunk 5 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5085]: will fix

File: usr/lib/common/mech_openssl.c (hunk 6 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5202]: Yes, in p11sak, we do print out the OpenSSL errors at the places where they happend.
But for the OCK libraries we don't. Instead, the API layer will print out all collected OpenSSL errors to the trace (it trace is enabled) in the END_OPENSSL_LIBCTX macro that ends every API function. 
So no need to do it here.

File: usr/lib/common/mech_openssl.c [hunk 7]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+    CK_ATTRIBUTE *prime1 = NULL, *prime2 = NULL;
+    BN_CTX *bn_ctx;
+    BIGNUM *n, *e, *p, *q, *d;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+
+    bn_ctx = BN_CTX_secure_new();
+    if (bn_ctx == NULL) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5011]: error in BN_CTX_secure_new()

File: usr/lib/common/mech_openssl.c [hunk 8]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_2,
+                                          &prime2);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_2\n");
+        goto done;
+    }
+
+    q = BN_CTX_get(bn_ctx);
+    if (q == NULL ||
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, q) == NULL) {
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5076]: prime2->pValue   and prime2->ulValueLen

File: usr/lib/common/mech_openssl.c [hunk 9]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    /*
+     * we use hardcoded hash so that migrating between versions that use
+     * different hash doesn't provide a Bleichenbacher oracle:
+     * if the attacker can see that different versions return different
+     * messages for the same ciphertext, they'll know that the message is
+     * syntethically generated, which means that the padding check failed
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5319]: ... another "synthetically"

File: usr/lib/common/mech_rsa.c [hunk 10]
```diff
@@ -328,35 +341,75 @@ static CK_RV rsa_parse_block_type_2(CK_BYTE *in_data,
...
 
-    good = constant_time_is_zero(em[0]);
-    good &= constant_time_eq(em[1], 2);
+    /*
+     * max message size is the size of the modulus size minus 2 bytes for
+     * version and padding type and a minimum of 8 bytes padding
+     */
+    len_mask = max_sep_offset = in_data_len - 2 - 8;
+    /*
+     * we want a mask so lets propagate the high bit to all positions less
```
Comment (jschmidb) [inline usr/lib/common/mech_rsa.c:376]: let's

File: usr/lib/common/mech_openssl.c (hunk 7 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5011]: fixed

File: usr/lib/common/mech_openssl.c (hunk 8 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5076]: good catch!  Fixed.

File: usr/lib/common/mech_rsa.c (hunk 10 above)
Comment (ifranzki) [reply usr/lib/common/mech_rsa.c:376]: fixed

File: usr/lib/common/mech_openssl.c [hunk 11]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime2\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(q, BN_FLG_CONSTTIME);
+
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL) {
+        TRACE_ERROR("BN_CTX_get failed to get d\n");
+        rc =  CKR_FUNCTION_FAILED;
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5086]: 2 spaces after rc = 
(also below)

File: usr/lib/common/mech_openssl.c [hunk 12]
```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
...
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_MODULUS,
+                                          &modulus);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        return rc;
+    }
+
+    buf = calloc(1, modulus->ulValueLen);
+    if (buf == NULL) {
+        TRACE_ERROR("Failed to allocate a buffer for private exponent\n");
```
Comment (jschmidb) [inline usr/lib/common/mech_openssl.c:5157]: buffer for modulus

Comment (jschmidb) [reply usr/lib/common/mech_openssl.c:5157]: Ah ok ... it's in fact the private exponent, which has the same length as the modulus ...

File: usr/lib/ep11_stdll/ep11_specific.c [hunk 13]
```diff
@@ -9596,10 +9596,12 @@ CK_RV ep11tok_decrypt_final(STDLL_TokData_t * tokdata, SESSION * session,
     rc = constant_time_select(constant_time_eq(rc, CKR_OK),
                               ep11_error_to_pkcs11_error(rc, session),
                               rc);
-    if (rc != CKR_OK) {
-        TRACE_ERROR("%s rc=0x%lx\n", __func__, rc);
-    } else {
-        TRACE_INFO("%s rc=0x%lx\n", __func__, rc);
+    if (!is_rsa_mechanism(ctx->mech.mechanism)) {
+        if (rc != CKR_OK) {
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/ep11_stdll/ep11_specific.c:9600]: This branch is not dependent on the data or the previous return code (which itself might be dependent on the data), so no need to do this in constant time.

File: usr/lib/common/mech_openssl.c (hunk 11 above)
Comment (ifranzki) [reply usr/lib/common/mech_openssl.c:5086]: fixed (3 places)

File: COPYRIGHTS [hunk 14]
```diff
@@ -12,19 +12,29 @@ For code originating from OpenSSL:
...
  * - usr/lib/common/constant_time.h: Copied unchanged from OpenSSL from
-     include/internal/constant_time.h
+ *   include/internal/constant_time.h
  * - The implementation of function rsa_parse_block_type_2() in 
  *   usr/lib/common/mech_rsa.c is copied from OpenSSL's function
  *   ossl_rsa_padding_check_PKCS1_type_2() in crypto/rsa/rsa_pk1.c
  *   and is slightly modified to fit to the OpenCryptoki environment.
  *   See comment in function rsa_parse_block_type_2() for a list of changes.
+ * - The implementation of function openssl_specific_rsa_derive_kdk() in
+ *   usr/lib/common/mech_openssl.c is copiled from OpenSSL's function
```
Comment (jschmidb) [inline COPYRIGHTS:22]: is copied from

File: COPYRIGHTS [hunk 15]
```diff
@@ -12,19 +12,29 @@ For code originating from OpenSSL:
...
+ *   derive_kdk() in crypto/rsa/rsa_ossl.c and is slightly modified to fit to 
+ *   the OpenCryptoki environment. See comment in function 
+ *   openssl_specific_rsa_derive_kdk() for a list of changes.
+ * - The implementation of function openssl_specific_rsa_prf() in
+ *   usr/lib/common/mech_openssl.c is copiled from OpenSSL's function
```
Comment (jschmidb) [inline COPYRIGHTS:27]: also here

**Review by jschmidb — APPROVED** [review]
looks good besides of the two last typos ...

File: COPYRIGHTS (hunk 14 above)
Comment (ifranzki) [reply COPYRIGHTS:22]: fixed

File: COPYRIGHTS (hunk 15 above)
Comment (ifranzki) [reply COPYRIGHTS:27]: fixed

**Pull request merged**